---
title: "Finemapping with missing data"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{R include=FALSE}
log.dmvnorm <- function(x, mu = rep(0, length(x)), S = diag(1, length(x)) ){
  #returns log of density of MV-Normal(mean = mu, var = S) at x 
  K = length(mu)
  stopifnot(all(dim(S) == K))
  stopifnot(length(x) == K)
  chol.S = chol(S) #Cholesky decomposition
  log.det = 2*sum(log(diag(chol.S))) #log of det(S)
  inv.chol.S = solve(t(chol.S)) #inverse of cholesky^T
  return(-K/2*log(2*pi) - 0.5*(log.det + crossprod(inv.chol.S %*% (x-mu))))
}
```

```{R}
# SIMULATING GWAS RESULTS FROM p SNPS WITH GIVEN CORRELATION MATRIX

### INPUT
p = 16
lambda.true = c(0.12, 0.15, 0.11, rep(0,p-3))
f = runif(p, 0.1, 0.5)
tau = 0.1

A = matrix(runif(p^2), ncol=p) 
B = t(A) %*% A
R = cov2cor(B)

n = floor(p*seq(1000, 2000, length.out = p))
N = matrix(NA, p, p)
for (i in 1:p) {
  for (j in 1:p) {
    N[i,j] = min(n[i], n[j])
  }
}

### COMPUTATION

M = matrix(NA, p, p)
for (i in 1:p) {
  for (j in 1:p) {
    M[i,j] = N[i,j]/sqrt(n[i]*n[j])
  }
}

# marginal effects and their SEs
sc = sqrt(2*f*(1-f))
lambda.sc = sc*lambda.true
beta.sc = R %*% lambda.sc
beta = beta.sc/sc
SE.b.sc = c(sqrt( (1-beta^2*2*f*(1-f))/n ))
S = diag(SE.b.sc)
Sigma = S %*% (R*M) %*% S


#matrix encoding causal configurations
C = matrix(NA, 2^p, p)
for (i in 1:p) {
  C[,i] = rep(c(rep(1,2^(i-1)), rep(0,2^(i-1))), 2^p/(2^i))
}
for (i in rev(1:2^p)) {
  if(sum(C[i,])>4 | sum(C[i,])==0) {
    C = C[-i,]
  }
}
true = 2510 # true config = first three SNPs causal
```

```{R}
# simulating one set of beta estimates from their joint distribution
library(MASS)
betahat.sc = mvrnorm(n = 1, mu = beta.sc, Sigma = Sigma)

log.bf = rep(NA, nrow(C))
for (i in 1:nrow(C)) {
  log.bf[i] = log.dmvnorm(betahat.sc, mu = rep(0,p), S = Sigma + R %*% diag(tau^2*C[i,]) %*% R ) - 
    log.dmvnorm(betahat.sc, mu = rep(0,p), S = Sigma)
}
bf = exp(log.bf - max(log.bf))
pk = rep(NA, 4)
for (k in 1:4) {
  pk[k] = choose(p, k)*(1/p)^k*((p-1)/p)^(p-k)
}



# unnormalized posterior for each config 
p_ = rep(NA, nrow(C))
for (i in 1:nrow(C)) {
  k = sum(C[i,])
  p_[i] = bf[i]*pk[k]/choose(p,k)
}

# normalized posterior
posterior.config = (p_)/sum(p_)
plot(posterior.config)

# top config
C[which.max(posterior.config),]

# posterior inclusion probability for each SNP
post.ip = rep(0, p)
for (snp in 1:p) {
  for (config in 1:nrow(C)) {
    if(C[config, snp]==1) {
      post.ip[snp] = post.ip[snp] + posterior.config[config]
    }
  }
}
plot(post.ip)
plot(bf)
posterior.config[true]
```


1. set threshold range to e.g. 0.4 - 0.9
2. compute, for each SNP, its total r^2*n_overlap, and check the other SNPs' contribution to this sum
3. define each SNP's "supporter set" of SNPs that cover at least the threshold % of total r^2n
4. to evaluate the BF and posterior of a causal config, consider the union of the causal SNPs' supporters
5. plot the true causal config's posterior (or the PIPs of true causal variants) against threshold used


```{R}
### FINDING SUPPORTERS FOR SNPS
r2n.sums = diag(R^2%*%N)
r2n = R^2 * N

#MI = -0.5*log(1-(R*M)^2) # problem: inf on diagonal
MI = R*(1-M)
MI.sums = rowSums(MI)

# info = matrix(0, p, p)
# for (i in 1:p) {
#   for (j in 1:p) {
#     if (N[j,j]>N[i,j]) {
#       info[i,j] = N[j,j]-N[i,j]
#     } else {
#       info[i,j] = N[j,j]-N[i,i]
#      }
#   }
# }
# info = R^2 * info
# info = info + abs(min(info))
# 
# info.sums = rowSums(info)

thresh = seq(0.6, 1.0, length.out=20)
support = array(0, c(p,p,length(thresh)))
aggregates = matrix(NA, length(thresh), p)

for (t in 1:length(thresh)) {
  thr = thresh[t]
  for (snp in 1:p) {
    aggr = 0
    i = p
    while(aggr<thr & i!=0) {
      nextbest = which(rank(MI[snp,])==i)
      support[snp, nextbest, t] = 1
      aggr = aggr + MI[snp,nextbest]/MI.sums[snp]
      i = i-1
    }
    aggregates[t, snp] = aggr
  }
}

for (t in 1:length(thresh)) {
  diag(support[,,t]) = rep(1, p)
}


### original r2n support
# for (t in 1:length(thresh)) {
#   thr = thresh[t]
#   for (snp in 1:p) {
#     aggr = 0
#     i = p
#     while(aggr<thr & i!=0) {
#       nextbest = which(rank(r2n[snp,])==i)
#       support[snp, nextbest, t] = 1
#       aggr = aggr + r2n[snp,nextbest]/r2n.sums[snp]
#       i = i-1
#     }
#     aggregates[t, snp] = aggr
#   }
# }
### 

### CALCULATING POSTERIORS FOR CONFIGS WITH THEIR SUPPORT

# prior
pk = rep(NA, 4)
for (k in 1:4) {
  pk[k] = choose(p, k)*(1/p)^k*((p-1)/p)^(p-k)
}

niter = 50
res = array(NA,c(niter, length(thresh), 4)) # true config posterior + PIPs for 3 causal SNPs

for (iter in 1:niter) { 
  betahat.sc = mvrnorm(n = 1, mu = beta.sc, Sigma = Sigma)
  for (t in 1:length(thresh)) {
    thr = thresh[t]

    log.bf = rep(NA, nrow(C))
    for (i in 1:nrow(C)) {
      
      if (is.vector(support[which(C[i, ] == 1), , t])) {
        conf.sup = which(support[which(C[i, ] == 1), , t]>0)
      } else {
        conf.sup = which(colSums(support[which(C[i,]==1),,t])>0 ) 
      }
      
      log.bf[i] = log.dmvnorm(betahat.sc[conf.sup], mu = rep(0,length(conf.sup)), 
                              S = (Sigma + R %*% diag(tau^2*C[i,]) %*% R)[conf.sup,conf.sup] ) - 
        log.dmvnorm(betahat.sc[conf.sup], mu = rep(0,length(conf.sup)), S = Sigma[conf.sup,conf.sup])
    }
    bf = exp(log.bf - max(log.bf))
    
    # unnormalized posterior for each config 
    p_ = rep(NA, nrow(C))
    for (i in 1:nrow(C)) {
      k = sum(C[i,])
      p_[i] = bf[i]*pk[k]/choose(p,k)
    }
    
    # normalized posterior
    posterior.config = (p_)/sum(p_)
    
    # posterior for true config (first 3 SNPs causal)
    posterior.config787 = (p_[787])/sum(p_)
    
    #PIPs
    post.ip = rep(0, p)
    for (snp in 1:p) {
      for (config in 1:nrow(C)) {
        if(C[config, snp]==1) {
          post.ip[snp] =  post.ip[snp] + posterior.config[config] #post.ip[snp] + p_[config]
        }
      }
    }
    res[iter, t, 1] = posterior.config[true]
    res[iter, t, 2:4] = post.ip[1:3]
  }
}
```

```{R}
dimprop = rep(NA, length(thresh))
for (t in 1:length(thresh)) {
  dimprop[t] = sum(support[,,t])/p^2
}

plot(dimprop, apply(res[,,1], 2, mean), type="l", lwd=2,  xlab="% of data used", ylab = "posterior of true config" )
#plot( apply(res[,,1], 2, mean), type="l", lwd=2, ylab = "posterior of true config", ylim=c(0,1) )

plot(dimprop, apply(res[,,2], 2, mean), type="l", lwd=2, xlab="% of data used", ylab = "PIP", ylim=c(0.0, 0.98))
legend("bottomright", c("causal 1", "causal 2", "causal 3"), col=c(1,2,3), lwd=2)
# points(thresh,apply(res[,,2], 2, mean), lwd=2, col=1)
# points(thresh,apply(res[,,3], 2, mean), lwd=2, col=2)
# points(thresh,apply(res[,,4], 2, mean), lwd=2, col=3)
lines(dimprop,apply(res[,,3], 2, mean), lwd=2, col=2)
lines(dimprop,apply(res[,,4], 2, mean), lwd=2, col=3)

```

